torch/csrc/jit/passes/shape_analysis.cpp

```diff
@@ -107,10 +107,9 @@ bool mergeTypes(ArrayRef<Value*> lhs, ArrayRef<Value*> rhs, ArrayRef<Value*> out
   bool changed = false;
   for(size_t i = 0; i < lhs.size(); ++i) {
     auto old_output_type = outputs[i]->type();
-    if(*lhs[i]->type() == *rhs[i]->type()) {
-      outputs[i]->setType(lhs[i]->type());
-    } else {
-      outputs[i]->setType(DynamicType::get());
+    auto new_type = unifyTypes(lhs[i]->type(), rhs[i]->type());
+    if (new_type) {
```

Comment:
Shouldn't we assert this has succeeded? We need to set the type somehow, but if it fails, then you have incompatible types depending on the trip count.

Reply:
Although I would expect an error to be thrown earlier in compilation than here